tests/trainer/test_trainer.py

```diff
@@ -1789,3 +1693,76 @@ def training_step(self, batch, batch_idx):
     model = TestModel()
     trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, gpus=1)
     trainer.fit(model, train_data)
+
+
+def test_train_loop_system(tmpdir):
+    """
+    Test the following methods are called in the order in automatic optimization.
+    1. training_step
+    2. zero_grad (skip when accumulate gradients)
+    3. backward
+    """
+    called_methods = []
+
+    trainer_options = dict(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        limit_train_batches=3,
+        limit_val_batches=1,
+        limit_test_batches=1,
+        progress_bar_refresh_rate=0,
+    )
+
+    class TestOptimizer(SGD):
+        def zero_grad(self, *args, **kwargs):
+            called_methods.append(inspect.currentframe().f_code.co_name)
```

Comment:
You can set the name yourself, no need to use the current frame

```suggestion
            called_methods.append("zero_grad")
```

same for the other occurrences